### PR TITLE
Polars: from_product() implementation

### PR DESCRIPTION
**Pull request by rjavadi**
## Added
Feature: Speed up `from_product()` through using Polars cross-join and filtering.
Test: Added to `test_constraint_polars.py`

### CODE REVIEW

**Comment by Scienfitz** [comment]
@rjavadi just a reminder, I think you already did this but because one can easily forget: make sure to base these branches on `dev/polars` and not `main`

File: baybe/searchspace/discrete.py
Comment (AVHopp) [inline baybe/searchspace/discrete.py:742]: Since we are in a discrete sub space, can it even happen that we have non-discrete parameters? Or is this check just to be sure?

File: pyproject.toml
Comment (AVHopp) [inline pyproject.toml:48]: Can you comment on why this is necessary?

**Review by AVHopp — APPROVED** [review]
Looking good to me 😄

**Comment by Scienfitz** [comment]
I rebased this PR onto `dev/polars` after merging #305

File: pyproject.toml
Comment (rjavadi) [reply pyproject.toml:48]: Core tests didn't pass due to missing dependency.

File: baybe/searchspace/discrete.py
Comment (rjavadi) [reply baybe/searchspace/discrete.py:742]: I didn't add this line, I only renamed the function  `_apply_constraint_filter` to ` _apply_pandas_constraint_filter` and it was part of `_apply_constraint_filter`.

File: pyproject.toml
Comment (Scienfitz) [reply pyproject.toml:48]: @AdrianSosic does this relate to the one thing you found at some time? I remember you saying that there was some dep installed/pinned only in fulltest and not core

File: baybe/searchspace/discrete.py
Comment (Scienfitz) [reply baybe/searchspace/discrete.py:742]: this check should stay here in analogy to what happens in the pandas version `parameter_cartesian_prod_to_df`
However the line should look more like `discrete_parameters = [p for p in parameters if p.is_discrete]`

File: baybe/searchspace/discrete.py
Comment (Scienfitz) [inline baybe/searchspace/discrete.py:761]: please rename to make clear this is the pandas version similar to the versions of constraint applicators

File: tests/test_constraints_polars.py
Comment (Scienfitz) [inline tests/test_constraints_polars.py:190]:
```suggestion
    assert len(pol_df.to_pandas()) == len(pd_df), "Polars and Pandas dataframes have different length."
```

File: baybe/searchspace/discrete.py
Comment (rjavadi) [reply baybe/searchspace/discrete.py]: Changing this line causes mypy error since I'm using `p.values`

Comment (Scienfitz) [reply baybe/searchspace/discrete.py:742]: you can ignore this error with a ignore statement because in this case mypy is wrong, this is also done in `parameter_cartesian_prod_to_df `

Comment (AdrianSosic) [reply baybe/searchspace/discrete.py]: @Scienfitz: I'm not sure if we've already had this conversation but: why do we actually introduce this property when the only real effect is that we lose the ability to typeguard?

File: pyproject.toml
Comment (AdrianSosic) [reply pyproject.toml:48]: I think what you are referring to is a different issue that was related to mixing our dependency groups. See commit 227182d7ee3e2e28ed2254292e102f02bd7fba64 (message: `Remove "lint" from "test" dependency group`). But that had nothing to do with the current one. Can you confirm?

I think the `pyarrow` here is probably required by some special `polars` feature, perhaps the `streaming`? Does someone of you know, perhaps? But anyway, I think it should thus be installed as a secondary dependency through `polars`, i.e. by replacing `polars` with `polars[pyarrow]` in our toml file

File: baybe/searchspace/discrete.py
Comment (AdrianSosic) [inline baybe/searchspace/discrete.py:263]:
```suggestion
        # Check if dataframe is empty to prevent Polars error
```

File: baybe/searchspace/discrete.py
Comment (AdrianSosic) [inline baybe/searchspace/discrete.py:264]: isn't a simple `len` sufficient here?

File: baybe/searchspace/discrete.py
Comment (AdrianSosic) [inline baybe/searchspace/discrete.py:265]: why is this if/else block needed in the first place? `pl.DataFrame().to_pandas()` works for me without problems. Am I overlooking something?

File: baybe/searchspace/discrete.py
Comment (AdrianSosic) [inline baybe/searchspace/discrete.py:270]: Now this applies the already used constraints **again**! This needs a cleverer logic

File: baybe/searchspace/discrete.py
Comment (AdrianSosic) [inline baybe/searchspace/discrete.py:258]: Not sure if we've decided to add it in this PR already, but the optionality of polars is missing

File: baybe/searchspace/discrete.py
Comment (AdrianSosic) [inline baybe/searchspace/discrete.py:682]: The inplace/return logic is not very consistent in both methods. Both versions currently operate inplace, but the pandas version returns `None` (btw: return type hint missing) while the polars version still returns the modified dataframe. I think this should be handled identically in both methods. Perhaps we should just always return the dataframe (no matter if modified inplace or not) and also add a boolean `inplace` argument to control it?

File: baybe/searchspace/discrete.py
Comment (AdrianSosic) [inline baybe/searchspace/discrete.py:761]: type should actually be `Sequence` (also in polars version)

File: baybe/searchspace/discrete.py
Comment (AdrianSosic) [inline baybe/searchspace/discrete.py:732]: info about polars is missing

File: baybe/searchspace/discrete.py
Comment (AdrianSosic) [inline baybe/searchspace/discrete.py:756]: I'm still not very clear about this `force_parallel` option, especially since there is also an `allow_parallel` option. Can we again summarize what is the intention here and why not omit it / use the other option? In the end, I think the code at least deserves a comment, but perhaps let's first clarify here

File: tests/test_constraints_polars.py
Comment (AdrianSosic) [inline tests/test_constraints_polars.py:190]: why not also test for equality here?

**Review by AdrianSosic — CHANGES_REQUESTED** [review]

File: baybe/searchspace/discrete.py
Comment (Scienfitz) [reply baybe/searchspace/discrete.py]: i think you yourself said sometime ago that we need to get rid of many `isinstance` calls that were there previously and I agree, the properties are the perfect way to do that, I dont think we should build our code to work aorund limitations of mypy

in this concrete case here nothing new has been introduced, the applications should just mimic what happens in the preexisting pandas constraint application function

File: baybe/searchspace/discrete.py
Comment (Scienfitz) [reply baybe/searchspace/discrete.py:270]: yes, let me redesign it in th next PR where optionality is included because the lines here will be affected

File: baybe/searchspace/discrete.py
Comment (AdrianSosic) [reply baybe/searchspace/discrete.py]: alright 👍🏼 in fact, perhaps we can solve the issue from a different angle in the future by
* upgrading mypy
* using some other checker such as basedmypy / pyright / etc....

But out of scope for this PR

File: baybe/searchspace/discrete.py
Comment (rjavadi) [reply baybe/searchspace/discrete.py:265]: In this way you're first creating a Polars dataframe and then do a conversion.
Since you were concerned about possible performance overhead, I wanted to avoid it (however it's negligible here)

File: baybe/searchspace/discrete.py
Comment (rjavadi) [reply baybe/searchspace/discrete.py:682]: @Scienfitz

File: baybe/searchspace/discrete.py
Comment (rjavadi) [reply baybe/searchspace/discrete.py:761]: Fixed

File: baybe/searchspace/discrete.py
Comment (rjavadi) [reply baybe/searchspace/discrete.py:264]: Right, fixed.

File: baybe/searchspace/discrete.py
Comment (rjavadi) [reply baybe/searchspace/discrete.py:732]: Fixed.

File: baybe/searchspace/discrete.py
Comment (rjavadi) [reply baybe/searchspace/discrete.py:756]: Using `allow_parallel` doesn't guarantee parallelization. 
The optimizer may still choose not to parallelize if it deems it unnecessary or potentially detrimental. `force_parallel`  overrides the optimizer's decision and ensures parallel execution

File: tests/test_constraints_polars.py
Comment (rjavadi) [reply tests/test_constraints_polars.py:190]: Added.

File: pyproject.toml
Comment (rjavadi) [reply pyproject.toml:48]: It was required for `join`, but I don't know exactly where.

Comment (AVHopp) [reply pyproject.toml:48]: Ok, so can we then resolve this? Seems like this is then something we need to include ad have no choice, right?

Comment (AdrianSosic) [reply pyproject.toml:48]: No, for two reasons:
* We should understand why/if pyarrow is really needed. For that, someone needs to conduct a test in an environment where it is **not** installed and determine the operation where it is needed.
* If it turns out it is really needed for polars, then we should change it into a polars secondary dependency, instead of a primary one

File: baybe/searchspace/discrete.py
Comment (AdrianSosic) [reply baybe/searchspace/discrete.py:265]: That was not my point, I just wanted to demonstrate that converting an empty polars dataframe to pandas causes no problems. So why do you then have this if/else block in the first case? The first case simply covers the case that you have an empty dataframe, and I do not see the reason why it should not be possible to simply call `df = pl_df.to_pandas()` in both cases?

File: baybe/searchspace/discrete.py
Comment (AdrianSosic) [reply baybe/searchspace/discrete.py:682]: Yes, I think @Scienfitz can take over from here

File: tests/test_constraints_polars.py
Comment (AdrianSosic) [reply tests/test_constraints_polars.py:190]: thx, but I think the previous assert can then go (since implied by the equality check)

File: baybe/searchspace/discrete.py
Comment (rjavadi) [reply baybe/searchspace/discrete.py]: I remember that there was a problem with empty Polars dataframes, but you're right, I no longer see related errors after running tests, so I remove the redundant check.

**TL;DR: Fixed**

File: baybe/searchspace/discrete.py
Comment (Scienfitz) [reply baybe/searchspace/discrete.py:679]: @rjavadi can the polars filtering even be done `inplace`? If not this request is obsolete (unless you want to add unnecessary copying to the pandas version @AdrianSosic )

File: baybe/searchspace/discrete.py
Comment (Scienfitz) [reply baybe/searchspace/discrete.py:753]: it seems to me that `allow_parallel` is the way to go then, is there any reason why you opted for `force_parallel` @rjavadi  ?

File: tests/test_constraints_polars.py
Comment (Scienfitz) [reply tests/test_constraints_polars.py]: removed

File: baybe/searchspace/discrete.py
Comment (AdrianSosic) [reply baybe/searchspace/discrete.py:753]: I'd also prefer to go with the least invasive option (since we are not sure about potential side effects)

File: baybe/searchspace/discrete.py
Comment (AdrianSosic) [reply baybe/searchspace/discrete.py:679]: Can't answer the first question since I don't know, but I think it makes sense to at least have the same output, regardless of whether we (can) do inplace or not.

File: baybe/searchspace/discrete.py
Comment (rjavadi) [reply baybe/searchspace/discrete.py:753]: I explained in the last [here](https://github.com/emdgroup/baybe/discussions/267) why I used `force_parallel`.

File: baybe/searchspace/discrete.py
Comment (rjavadi) [reply baybe/searchspace/discrete.py:679]: @Scienfitz Unfortunately, the `filter` function doesn't provide an option for in-place filtering.

File: baybe/searchspace/discrete.py
Comment (Scienfitz) [reply baybe/searchspace/discrete.py:753]: hmm comparing the risks here, it seems `force_parallel` can do speedup for large spaces. Given the prelim investigation it also seems that it was necessary to set this flag as without it polars did not opt for the parallelization. `allow_parallel` on the other hand seems the pick for small spaces, as then the parallelization is turned off if detrimental. But since the small space regime doesnt really matter (even if there was a difference of 10x computation time depending on which flag to choose) then I would leave `force_parallel`, @AdrianSosic lets do that in the interest of moving forward we can change it anytime again if issues would be encountered

File: baybe/searchspace/discrete.py
Comment (Scienfitz) [reply baybe/searchspace/discrete.py:679]: conceptually i guess it doesnt make sense to taLk about inplace filtering for a frame thats not even fully in nemory like a lazyframe,c orrespondingly the pandas utilities would have to oerpate on df copies and return that copy to be the same as polars filter, but I'd find that bad

Comment (AdrianSosic) [reply baybe/searchspace/discrete.py:679]: Yes, that's true. I'll leave it up to you then to decide if we can nevertheless return the inplace-modified dataframe. But personally, I'm not a big fan of these types of non-return functions that actually modify something that you want to use in the next step, because they don't allow you to chain operations. For example, every time I use our `add_fake_results` (which works in the same way), I run into the error that my next calling function tells me that `None` doesn't have certain attributes – because I forwarded that empty output to the next processing layer. And then it always requires me to create unnecessary temporary variables. For example
```python
# does not work when `None` is returned
next_operation(add_fake_results(pd.Dataframe(...)))

# current necessary workaround
tmp = pd.DataFrame(...)
add_fake_results(tmp)
next_operation(tmp)
```

I have to say that I quite like `pytorch`s approach here, which cleanly distinguishes in-place vs copy operation in terms of the trailing underscore (e.g. `add` vs `add_`) but the in-place version STILL returns the object. So you don't need to modify your code when switching from one to the other

File: baybe/searchspace/discrete.py
Comment (AdrianSosic) [reply baybe/searchspace/discrete.py:753]: Alright 👍🏼

File: baybe/searchspace/discrete.py
Comment (Scienfitz) [reply baybe/searchspace/discrete.py:679]: ok then returning simply the object even if altered inplace solves your request? i dont have any reason why it needs noreturn so we can easily alter that (or give it a flag anyway but always return it)

Comment (Scienfitz) [reply baybe/searchspace/discrete.py:679]: I'll put this on the list to do in the integration PR next up, unresolve if the proposed change is not enough

File: pyproject.toml
Comment (Scienfitz) [reply pyproject.toml:48]: Polars `.to_pandas` seems to require `pyarrow`, see [here](https://docs.pola.rs/api/python/stable/reference/dataframe/api/polars.DataFrame.to_pandas.html)
image
seems pretty unavoidable to me

Comment (Scienfitz) [reply pyproject.toml:48]: ok, seems to be avoidable by changing the corresponding searchspace creation step from
```python
pl_df = lazy_df.collect(streaming=True)
df = pl_df.to_pandas()
```
to 
```python
df_dict = lazy_df.collect(streaming=True).to_dict()
df = pd.DataFrame.from_dict(df_dict)
```
or
```python
df_dicts = lazy_df.collect(streaming=True).to_dicts()
df = pd.DataFrame.from_records(df_dicts)
```
(Haven't tested any impact on speed or memory)
Would that be fine @AdrianSosic ? If so we can avoid pyarrow dep

**Review by Scienfitz — APPROVED** [review]

File: pyproject.toml
Comment (AdrianSosic) [reply pyproject.toml]: Hey @Scienfitz, I would not cut down here on code on performance. For me, it's absolutely OK to have pyarrow as a transitive dependency – it was only important to me to be certain that we really need it. So keep the code as is, but please change it into a proper transitive dep by moving it to the polars entry 👍🏼

Comment (Scienfitz) [reply pyproject.toml]: im confused, its pretty clear that we dont need it, so why do you want to use it now? We dotn know anything about performance

**Review by AdrianSosic — APPROVED** [review]

File: pyproject.toml
Comment (AdrianSosic) [reply pyproject.toml]: The point is here that you go via dict conversion, which is definitely suboptimal. The pyarrow approach does it directly, even avoiding copies (see documentation). So it's clear that the dict approach does not require pyarrow since you basically deserialize your data, but that's also exactly where the downside lies

Comment (AdrianSosic) [reply pyproject.toml]: In a sense, pyarrow gives you a direct polars <-> pandas converter, while your solution makes the de-/serialization detour

Comment (Scienfitz) [reply pyproject.toml]: so to be clear you do *not* want to avoid the additional dependency (even if its introduced via `polars[pyarrow]`)? Because thats the vibe i got from previous comments

Comment (AdrianSosic) [reply pyproject.toml]: What? No, I *do want* to have it as `polars[pyarrow]`. Btw, another reason agains going via dicts, you can lose type information (also mentioned in the docs). 

And here you have the speed black on white, pretty convincing I think :D 
Speed: ~1s vs ~125s
Benchmark script:
```python
from time import perf_counter
import pandas as pd
import polars as pl
import numpy as np

df = pl.DataFrame(np.random.random((10000, 1000)))
t = perf_counter()
for _ in range(100):
    df.to_pandas()
    # pd.DataFrame.from_records(df.to_dicts())
print(perf_counter() - t)
```

Comment (Scienfitz) [reply pyproject.toml]: ok that is a pretty obv benefit, the confusions stems from this comment 
image
 which doesnt mention speed at all and sounds like `whether its essential`. Its clearly not essential (without the previously not mentioned speed requirement). 

In any case, I updated this PR and will merge it now, if theres any further discusison needed it can be done in #318 and subsequent

Comment (AdrianSosic) [reply pyproject.toml]: Yes, when I wrote that statement, I was not thinking of the possibility of hacky workarounds 😄 Great, thx for the fix

**Pull request merged**
